### DISCUSSION TITLE
Alpine static builds using github action built musl-cross-make toolchains

### DISCUSSION THREAD

**Discussion by userdocs** [aristocratos/btop#160] [Ideas]
Hello, first, thank your for the project.

I was recently made aware of this project and I noticed that you deploy Linux static binaries using Alpine and docker images from [muslcc/x86_64](https://hub.docker.com/r/muslcc/x86_64) that currently use `gcc version 10.2.1 20210227 (GCC)`

```bash
docker run -it muslcc/x86_64:aarch64-linux-musl ash -c 'gcc -v'
```

As this is similar to some build processes that I use I thought I might try to apply my my custom [musl-cross-make](https://github.com/userdocs/musl-cross-make) toolchains that are configured to use the latest available options from the parent project

```bash
CONFIG_SUB_REV = 888c8e3d5f7b
GCC_VER = 11-20211113
BINUTILS_VER = 2.37
MUSL_VER = git-master
GMP_VER = 6.2.1
MPC_VER = 1.2.1
MPFR_VER = 4.1.0

LINUX_VER = 5.15.2
```

These builds are created from a 100% github action based repo that currently uses triples of specific interest to me but as it's a matrix build I am sure it could be extended to use all [supported triples](https://git.zv.io/toolchains/musl-cross-make/-/blob/master/scripts/triples.txt) and provide comparable toolchains options.

So i build my own toolchains using github and they are provided as releases here - https://github.com/userdocs/musl-cross-make/releases/latest

I then used them to build btop - https://github.com/userdocs/btop-crossbuilds/releases/latest

![image](https://user-images.githubusercontent.com/16525024/142457216-5af9f153-bfa6-4865-811e-133a7cf30fb8.png)

One successful, full toolchain action, using all supported triples could provide the option to deploy a musl static binary using gcc 11 and provide you more control over the build and release than those docker images.

My builds in the examples are working, based on existing and demonstrable actions so i think the idea is a achievable with out too much drama. It was not too hard for to adapt my workflow to build your app.

**Comment by userdocs** [comment]
I thought i'd try the full 51 triples available and i got 50/51 

https://github.com/userdocs/musl-cross-make/runs/4256989016?check_suite_focus=true

The `riscv32-linux-musl` failure appears muslc related so i'll need to pick that up with the dev over at gitlab 

```
2021-11-18T22:21:38.3505460Z checking target system type... riscv32-linux-musl
2021-11-18T22:21:38.3507026Z ../src_musl/configure: unknown or unsupported target "riscv32-linux-musl"
2021-11-18T22:21:38.3519862Z make[1]: *** [Makefile:238: obj_musl/.lc_configured] Error 1
2021-11-18T22:21:38.3521213Z make[1]: Leaving directory '/root/musl-cross-make/build/local/riscv32-linux-musl'
2021-11-18T22:21:38.3548883Z make: *** [Makefile:194: all] Error 2
2021-11-18T22:21:38.3660523Z ##[error]Process completed with exit code 2.
```

But 50 latest toolchains from a github action is ok I think.

https://github.com/userdocs/musl-cross-make/releases/tag/11-2021

With 46/51 crossbuilds

https://github.com/userdocs/btop-crossbuilds/actions/runs/1478838385

https://github.com/userdocs/btop-crossbuilds/releases/tag/v1.1.1

**Reply by zv-io** [reply]
The `riscv32` target requires patches that must be applied only for that target, so they are "disabled" by default. I can show you how to enable them. That would essentially mean you have 51/51 targets.

**Reply by userdocs** [reply]
That would be really cool. I don't personally use or need it but i'd be happy to have successfully built the full suite and have them there as releases.

So what do I need to do?

btw, thanks for these toolchains, I find them really useful.

**Reply by zv-io** [reply]
This isn't documented (I've made a TODO for it), but there are two important things to know.

1. You cannot use `git-master` for musl if you want to apply any musl patches (it's a moving target). So I pin musl patches to a commit or specific official release. The `riscv32` functionality (not in upstream musl) is provided via a patchset and updated for `git-b76f37fd5625d038141b52184956fb4b7838e9a5` as of a few days ago. You'll find a directory called `patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5/`. Note further that there are 3 more patches I've not included here yet; I'm still testing them. All of this (`riscv32` target) is highly experimental. Since these patches were not active at the time of build, your 50 toolchains should be correct. However, consider:

1. These `riscv32` patches, if enabled, must not be included when building other targets. Imagine you started with `MUSL_VER = git-b76f37fd5625d038141b52184956fb4b7838e9a5` and built only the `riscv32-linux-musl` target. Now you want to build the others, so you must rename `patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5` temporarily and start with a pristine tree.

tl;dr: You got lucky in not including the `riscv32` patches in your build, and should build that target last after temporarily renaming the directory containing those patches while you build the rest of the toolchains. The latest patch set (in testing still, incomplete) is for `git-b76f37fd5625d038141b52184956fb4b7838e9a5`.

**Reply by userdocs** [reply]
ok, thanks for the info. I will try it but the site https://git.zv.io/toolchains/musl-cross-make seems to have gone missing (for me anyway) when it's back up i'll test it

Since i'm using a matrix I think I can apply the patch to that only when using `${{ matrix.arch_type }}` .

**Reply by zv-io** [reply]
It's down for backup and migration to new hardware; it might be a few hours.

**Reply by userdocs** [reply]
The site is back and before I try this can I just clarify something quickly?

So we have the patches here [musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5](https://git.zv.io/toolchains/musl-cross-make/-/tree/master/patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5)

My matrix build takes this template, edits for the matrix items and builds it

```
TARGET = MATRIX_TARGET < will be changed

CONFIG_SUB_REV = 888c8e3d5f7b
GCC_VER = 11-20211113
BINUTILS_VER = 2.37
MUSL_VER = git-master
GMP_VER = 6.2.1
MPC_VER = 1.2.1
MPFR_VER = 4.1.0

LINUX_VER = 5.15.2
```

So I can use this for 50 targets but when I want to build `risc32` I can just do this

```
TARGET = MATRIX_TARGET < will be changed

CONFIG_SUB_REV = 888c8e3d5f7b
GCC_VER = 11-20211113
BINUTILS_VER = 2.37
MUSL_VER = git-master < will be changed to musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5
GMP_VER = 6.2.1
MPC_VER = 1.2.1
MPFR_VER = 4.1.0

LINUX_VER = 5.15.2
```

the way I interpret your post is that when not building `risc32` I should rename the patch dir. To me this means all patches are `risc32` specific therefore equivalent to using `MUSL_VER = git-master` for all non `risc32` targets as no patches are applied either way, but potentially untested? Is a successful built enough to determine a positive test? 

If the I am not using any patches for non `risc32` items I can continue to use git master for 50 targets and just use the pin patch for the specific matrix item `risc32`, is this correct? (aside from the untested aspect)

**Reply by zv-io** [reply]
I would strongly recommend never using `git-master` unless you're trying to quickly test the latest version of musl. The musl patches that I have available now are *coincidentally* only for `riscv32` now, but there may be other patches you'd want.

**Reply by userdocs** [reply]
As far as i understand this:

> tl;dr: You got lucky in not including the riscv32 patches in your build, and should build that target last after **temporarily renaming the directory containing those patches while you build the rest of the toolchains**. The latest patch set (in testing still, incomplete) is for git-b76f37fd5625d038141b52184956fb4b7838e9a5.

When i set `MUSL_VER = git-b76f37fd5625d038141b52184956fb4b7838e9a5` it will use all the patches in the folder 

https://git.zv.io/toolchains/musl-cross-make/-/tree/master/patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5

So what i am confused about is, if i am to set this in my `config.mak`

```bash
MUSL_VER = git-b76f37fd5625d038141b52184956fb4b7838e9a5
```

But then rename the dir as to avoid using the patches

```
patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5
```

I am not using any of those patches? 

Am i missing something here about what you said?

I want to use the patches but i don't understand how if i rename the dir?

**Reply by zv-io** [reply]
The patches are applied only if the names match exactly. For example, `MUSL_VER = 1.2.2` looks for a directory called `patches/musl-1.2.2/`. For git commits, `patches/musl-git-FOO/`. If the directory does not exist, then it silently does not apply.

I usually just append `.bak` to the directory before building, then rename it back, and push that to the repo.

**Reply by userdocs** [reply]
I think i understand that part but the bit i am struggling with is understanding how to implement this

> These riscv32 patches, if enabled, must not be included when building other targets. Imagine you started with MUSL_VER = git-b76f37fd5625d038141b52184956fb4b7838e9a5 and built only the riscv32-linux-musl target. Now you want to build the others, so you must rename patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5 temporarily and start with a pristine tree.

So if i do those for the other 50 targets, there are no musl patches applied, and there are 14 patches here:

https://git.zv.io/toolchains/musl-cross-make/-/tree/master/patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5

7 are have `riscv` in the name so i assume these are specific to this target, but what about the other 7? do i need them? are all 14 `riscv` specifc?

I don't understand how using this

```
MUSL_VER = git-b76f37fd5625d038141b52184956fb4b7838e9a5
```

Provides access to any patches if I rename to folder for 50/51 targets.

Instead of rename the folder should I delete the patches 08-14

This is what I am trying to underhand. Which patches I need for musl, for all 51 targets,

**Reply by zv-io** [reply]
The current musl matches (`patches/musl-git-b76f37fd5625d038141b52184956fb4b7838e9a5/`) are currently:
* Only for `riscv32` and highly experimental
* Unique in that they will break other targets in subtle ways if they're applied
Normally the patches are non-disruptive in that they can safely be applied without undesirable side-effects. For now this is not true of this patch set.

The names (first 7 vs second 7) are not reliable. This entire patch set is for `riscv32` only. There are three additional patches that are not included yet because I haven't finished testing them.

**Because these 14 patches are all `riscv32`-specific, renaming is necessary for the other 50 targets.**

Only apply them for `riscv32-linux-musl`.

**Reply by userdocs** [reply]
Ok, I understand how it works now.  So I did that and I got all 51/51 https://github.com/userdocs/musl-cross-make/runs/4274867415?check_suite_focus=true

Since that particular commit is also the [latest in the master branch ](https://github.com/bminor/musl/commit/b76f37fd5625d038141b52184956fb4b7838e9a5) the outcome is effectively the same for these build as I still used `= git-master` for the other 50. In future ill look to see if the patches or non breaking but if not match the commit.

I [imported the repo over here](https://gitlab.com/userdocs/musl-cross-make) and i'll try to mimic the workflow and see if i can set up a like for like CI.

edit: fixed link

**Comment by aristocratos** [comment]
@userdocs 
Hi, the current release builds are scripted and built locally on my dev machine (for more control and quicker debugging) and I will swap over to use your toolchain for the these builds.

Changing the "Continuous Build Linux" workflow to use a musl gcc 11 toolchain seems like a great idea, I not sure if it really matters that much if 32-bit RISC can't be built, I'm gonna assume that anyone running that platform are likely gonna be comfortable compiling btop themselves if need be :)

I would like to ask @jan-guenter who setup the current workflow if he has any thoughts on the subject?

**Reply by aristocratos** [reply]
Also, I think if we copy over `README.md` and the `Makefile` to the `btop-completed` directory, we've got a working installer from the workflow.

**Reply by userdocs** [reply]
The workflow here has a larger matrix so I tried to build against that https://github.com/aristocratos/btop/blob/main/.github/workflows/continuous-build-linux.yml

I was mostly successful in this but if you are using more standard targets to actually release (14 linux targets?) https://github.com/aristocratos/btop/releases/tag/v1.1.1

You also do release a `btop-1.1.1-riscv32-linux-musl.tbz` build so i was not sure how important it was to achieve this

**Reply by aristocratos** [reply]
I think he vast majority of people is probably after the x86 and arm based builds, the other targets included in the release builds I mostly just threw in there, so no worries about removing any that are troublesome.

**Reply by userdocs** [reply]
If that is the case and since you are a small binary you have the option of full emulation builds as i mentioned to @jan-guenter 

https://github.com/userdocs/iperf3-static/blob/master/.github/workflows/multiarch_qemu_iperf3.yml

```
arch: [i386, amd64, arm32v6, arm32v7, arm64v8, ppc64le, s390x]
```

From this method (which is part of my matrix build) https://github.com/multiarch/qemu-user-static

Using these dockers:

https://hub.docker.com/r/s390x/alpine/
https://hub.docker.com/r/ppc64le/alpine/
https://hub.docker.com/r/arm64v8/alpine/
https://hub.docker.com/r/arm32v7/alpine/
https://hub.docker.com/r/arm32v6/alpine/
https://hub.docker.com/r/amd64/alpine/
https://hub.docker.com/r/i386/alpine/

The caveat with this method is the build time can be greatly increased but since you are a small binary taking about 2 minutes i don't think it really matters if it takes 5-10 minutes instead.

So this is another valid option but you lose many of the targets available from musl-cross

**Reply by userdocs** [reply]
on no, ignore me. we don't get access to gcc 11 this way and that is why we are using musl-cross. I forgot that part.

**Reply by aristocratos** [reply]
The build method for the release builds doesn't really matter that much since I will be running them locally, so will just have to change my current build script a bit to use your toolchain instead.

However the continuous build workflow I intend for to include as many architectures as possible.

**Comment by jan-guenter** [comment]
In general I think it's a good idea to upgrade gcc, however I have to wonder why neither Alpine nor muslcc have yet released a gcc 11 version. For the past releases there have always been some patches necessary to ensure a fully working toolchain. (see the [patches directory of musl-cross-make](https://github.com/richfelker/musl-cross-make/tree/master/patches))

@zv-io, is there a special reason why muslcc has not been built with gcc 11 yet? Are you aware of any known complications using gcc 11 and musl?

@userdocs, were you able to complete the gcc test suit with your toolchain builds?
On the first glance I see nothing wrong the builds. I'll take closer look tonight and try to run the gcc test suits myself.

@userdocs, is there a reason why your github action uses docker manually instead of `container.image` and then running the `steps` directly inside the container? Also what's the qemu installation used for?

@aristocratos 
> Also, I think if we copy over `README.md` and the `Makefile` to the `btop-completed` directory, we've got a working installer from the workflow.

Told you some time ago that you could automate the whole release process ;-)
Perhaps another make target like `package` or similar producing the installer archive would be beneficial for a simple automated package build.

**Reply by aristocratos** [reply]
> Told you some time ago that you could automate the whole release process ;-)
Perhaps another make target like package or similar producing the installer archive would be beneficial for a simple automated package build.

Yeah I know :)
I meant including the Makefile in the created artifacts for quickly installing "non-release" builds could be a good idea.
(Just need to add a check for a `./src` directory to suppress some warnings).

I still plan on building the releases locally and only including some of the most common targets, for more "exotic" targets I refer to the continuous build.

**Reply by userdocs** [reply]
[Musl.cc](http://musl.cc) is based on this repo https://git.zv.io/toolchains/musl-cross-make which is in turn an extension of the original project https://github.com/richfelker/musl-cross-make/issues/99 . This is what I use and this project has [updated to gcc 11](https://git.zv.io/toolchains/musl-cross-make/-/tree/master/patches) but not published them onto the site.

@jan-guenter I don't think I specified anything test suite specific but I am confident it would work

The reason my workflow has qemu is since it's a template workflow file I use for various projects but it's not specific to this project. It's just bootstrapping the host to allow emulation inside the dockers if needed, I do not use `container:` for a few reasons I'll list.

- I had some problems with releases and I was building hundreds of of [assets and artifacts a day over the course of a few days](https://github.com/ncipollo/release-action/discussions/144). This method was 100% consistent so I transitioned over.
- I can do stuff on the host before I use the docker container whereas the `container:` drops me inside it before I can use the host.
- I need or want to be able to do things on the host or docker interchangeably during the build process that using `container:` does not allow me to do.
- I can bootstrap multiple dockers with unique envs and call them at will. So I could also use https://hub.docker.com/r/arm64v8/alpine/ alongside the `x86_64` alpine edge image whilst using the host runner

From this projects point of view there may be no effective difference to the outcome of using the `container:` method. If that works for you that's fine by me. The template I use across various build projects I have give me more control over the workflow process, that's all.

As long as the toolchain exists somewhere to use I am sure it can be applied via a small tweak to your existing process. Mine is just more a [proof of concept it can be done
](https://github.com/userdocs/btop-crossbuilds/releases/tag/v1.1.1)

**Reply by jan-guenter** [reply]
@userdocs, don't get me wrong, I actually like the idea behind your template, and the capability to run different architecture containers is quite interesting.
The only thing I dislike about it is that the actual build instructions are "obfuscated" by all the docker exec clutter.
Could you give me an example workflow where you use the emulation feature of your workflow? Maybe we could turn it into a nicely parameterized and reusable GitHub action; I think something like that could be of interest to many people.
It also gets around a problem I often had with GitHub workflows where certain pre-built actions don't work inside some container images because the pre-built binaries, like node, used in those actions are built specifically for the ubuntu runners.

I knew that muslcc is based on musl-cross-make, but I wasn't aware of Zachs fork on GitLab since the muslcc webpage links to the original by Rich. Thanks for that info.

I will fork your musl-cross-make and try to add a verification step using the gcc test suits, later today. Good thing you already have qemu in the workflow, since the test need to run on the target platform :-D

**Reply by userdocs** [reply]
Here where I test built binaries by getting their version

https://github.com/userdocs/cmake-crossbuild/blob/main/.github/workflows/cmake-ninja-crossbuild.yml#L160-L163

And here, where the idea originated from, to cross build qt6 without needing to have or build a host version. This due to cmake not being able to use non target arch stuff during the build process and qt6 needing to use host native tools to compile.

https://github.com/userdocs/qt6-crossbuild/blob/main/.github/workflows/qt6_crossbuild.yml

So it kind of of just became part of my template.

**Reply by jan-guenter** [reply]
@userdocs Thanks for the details.
Just FYI: Even thou you cannot build host tools for the build process inside the same CMakeFile, a common technique is to create separate CMakeFiles for the host tools and adding them as cmake build steps to the main CMakeFile using the ExternalProject module which allows you to specify the platform or toolchain used for the external project.

**Reply by userdocs** [reply]
Thanks for the info though am not familiar enough with cmake to know what that means without further research. If you have a link i can read that would help me.

With QT6 I was just following their docs https://doc.qt.io/qt-6/qt6-buildsystem.html#cross-compiling-qt and it appears to a be a cmake + qt6 specific thing.

It turns out that since qemu emulation on the host works inside generic docker images via the multiarch bootstrapping command

```bash
sudo docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

So I was able to avoid that requirement as the host native tools were emulated for the build process. Since I tested official documented methods and my emulation method and I got identical binary outcomes so I was happy enough to use that shortcut.

Since I needed to configure the host before the container is spun up to do this on a runner, this method was merged together from another Alpine build method using [full emulation to build small binaries](https://github.com/userdocs/iperf3-static/blob/master/.github/workflows/multiarch_qemu_iperf3.yml) instead of cross building but this limits me to the officially supported alpine platforms.

So this particular approach is a hybrid of certain approaches to building / cross building static binaries on Alpine. Whilst i agree that the docker exec commands are not the prettiest solution I am ok with the trade off for the flexibility it provides me to boot strap as many dockers and targets as I need whilst having the host runner available to me during the build process.

Using the env file also provided some useful options like modifying the envs on the fly for specific needs, here making bionic use gcc-8

```
- name: Docker target - bionic - reconfigure build vars
  if: matrix.os_version_id == 'bionic'
  run: |
    echo "CC=${{ matrix.CHOST }}-gcc-8" >> env.custom
    echo "AR=${{ matrix.CHOST }}-gcc-ar-8" >> env.custom
    echo "CXX=${{ matrix.CHOST }}-g++-8" >> env.custom
```

I appreciate you saw the potential of some elements of the method through the clutter. Part of the docker command could; potentially be put in an env except for `-w /path` as I use that instead of cd as using cd would require to use `docker exec ... /bin/ash -c 'cd path'

```
env:
multiarch: docker exec
```

Ultimately I find this approach highly flexible and adaptable where the `container:` method, though convenient, was starting to hinder me.

**Reply by jan-guenter** [reply]
We get a little side-tracked here, and yes I'm aware I started it :sweat_smile:
Let's talk somewhere else about the stuff not related to btop++. You don't have a public email on your profile, so if you like write me an [email](mailto:jan.guenter@googlemail.com) or contact me on [Discord](https://discordapp.com/users/jellow#9966).

**Reply by zv-io** [reply]
> is there a special reason why muslcc has not been built with gcc 11 yet? Are you aware of any known complications using gcc 11 and musl?

I started preparing for a new release on 2021-11-15 with GCC 11. It takes a very long time to build and test these toolchains (~1500 of them in total, from GCC 5 to 11). So far 11 looks good, except that the `sh4` targets are not working correctly and I need to carve out some time to debug that.

If you want to test some of these out in the interim, I'd be happy to upload some of the "release candidates".

**Reply by jan-guenter** [reply]
@zv-io thanks for taking the time to answer our questions.
My comment was not at all meant as a criticism; sorry if I came across the wrong way. There is no hurry, at least not for me. I know very well how long it can take to get something released.
It is up to @aristocratos if he wants to use pre-release toolchains for his project.

**Reply by aristocratos** [reply]
Honestly, I see no issue in using @userdocs's toolchain, or is there some reasoning I'm missing for waiting for the future musl.cc gcc11 release?

**Reply by userdocs** [reply]
Whilst the toolchains are unmodified versions of the musl.cc build combinations, my toolchains are meant to be a rolling release using the most modern options as they become available. So that repo might not be best suited for a stable resource for the cross build files.

When musl.cc releases a more current version it does not get changed for a while so we could consider this a stable release in comparison but you might need a more modern option 6 months later and musl.cc might not release until 12 months later. This is a perfectly normal scenario to consider.

Since I have demonstrated an effective yet simple method to create release via an action what might be sensible is to build and host your own? I cannot say that using my repo will provide a stable option since I might break stuff with new releases.

**Reply by jan-guenter** [reply]
Me neither. @userdocs toolchains are built using Zachs version of musl-cross-make, which is used to build the muslcc toolchains, and as he said:
> So far 11 looks good, except that the `sh4` targets are not working correctly (...)

so I don't expect any major issues using it.

Alternatively you could take up Zachs offer of providing the RC versions of his toolchain builds, which could provide him with better feedback in case of errors than a 3rd party build.

As I said, it's up to you, @aristocratos.

**Reply by aristocratos** [reply]
@jan-guenter @zv-io
> Alternatively you could take up Zachs offer of providing the RC versions of his toolchain builds, which could provide him with better feedback in case of errors than a 3rd party build.

I'm absolutely for it if you guys are willing to help setup the workflow and and so on?

**Reply by jan-guenter** [reply]
We could use the workflow of @userdocs which already works fine with toolchain tar balls. Only the download URLs for the toolchains would needed to be changed. When the final release is out we can switch back to the docker approach or simply change the URLs to the release toolchains of muslcc.

@zv-io We would like to take you up on your offer to test the RC versions of your gcc 11 toolchains. Would you have some time could you upload the somewhere and let us know? Also let us know if the downloads will be available till the final release or if you should copy them somewhere else.

**Reply by zv-io** [reply]
GCC 11 has been officially released and made the default (homepage, Docker images) as of just now. See website for details. Earlier versions are still being built and will be uploaded in the next day or so.

**Reply by jan-guenter** [reply]
Thanks a bunch @zv-io!
@aristocratos since the docker images got updated, the current workflow will automatically pick up the new toolchains on the next build. No changes necessary.

**Reply by aristocratos** [reply]
@jan-guenter @zv-io 
Mostly builds fine but got 7 failed ones: https://github.com/aristocratos/btop/actions/runs/1493628817
With segfaults from `ld` on most of them.

**Reply by jan-guenter** [reply]
@zv-io I narrowed down the issue to the `-static` flag.
It is reproducible with a single cpp file with "empty" main:
```cpp
int main() { return 0; }
```
compiled with:
```bash
g++ -o test test.cpp -static
```
leads to the error, while `-static-libgcc` and `-static-libstdc++` work just fine.
Here is a strace and core dump I took from within a `muslcc/x86_64:or1k-linux-musl` container using the example above: [or1k_trace_and_dump.tar.gz](https://github.com/aristocratos/btop/files/7587695/or1k_trace_and_dump.tar.gz)

I'll try to build a debug version of the toolchain to get a useable stacktrace.

**Reply by zv-io** [reply]
This is very interesting. I wonder if it is related to another GCC bug I found; I will spend some time today bisecting GCC first and then binutils once I get more information.

**Reply by zv-io** [reply]
`or1k-linux-musl` was 100% unmodified; the `sh*` variants have a nasty patch that I'll need to review more closely.

**Reply by zv-io** [reply]
The file `patches/gcc-11-20211120/0004-static-pie.diff` triggers this issue.

**Reply by zv-io** [reply]
This is definitely a GCC regression and I'll triage it after work today. The good news is this isn't an unrelated issue :)

**Reply by userdocs** [reply]
I rebuilt the github toolchain [using current settings](https://raw.githubusercontent.com/userdocs/musl-cross-make/main/config.mak) and i did not get these failures as my results were consistent with the previous toolchain version:

https://github.com/userdocs/btop-crossbuilds/actions/runs/1495346456

47/51 but if we minus microblaze as it's not built here then I am at 47/49 with no seg faults using an identical configuration to your release.

I'm not sure what it means but i think the consistent outcome has some value?

**Comment by zv-io** [comment]
@aristocratos

> Honestly, I see no issue in using @userdocs's toolchain, or is there some reasoning I'm missing for waiting for the future musl.cc gcc11 release?

There should be little or no issue in using these toolchains for your project here. But please be warned, that if you promote them to other developers for use in their projects, that's potentially a lot of untested and experimental code making its way into the wild and there may be subtle breakage. So my recommendation is use it here if you need bleeding-edge, but don't promote it elsewhere.

The toolchains on musl.cc are built with very specific versions of core system utilities and libraries.

@userdocs 

> ... but you might need a more modern option 6 months later and musl.cc might not release until 12 months later. This is a perfectly normal scenario to consider.

Prior to this past spring, releases were done every 1-2 months. I want to get back into that cadence.

@jan-guenter 

> Alternatively you could take up Zachs offer of providing the RC versions of his toolchain builds, which could provide him with better feedback in case of errors than a 3rd party build.

This won't be necessary.

**Reply by jan-guenter** [reply]
Sorry, seems I was still typing when you posted your reply :-D

@aristocratos, if you want to use @userdocs toolchains I think you can safely merge his workflow since it seems to work fine. You could rename it and simply disable the docker workflow if you like to switch back at some point.

**Reply by zv-io** [reply]
The next set of GCC 11 toolchains should be available officially on musl.cc tomorrow if I can resolve these remaining `sh4` issues and nothing else arises. This means new Docker images and defaulting to GCC 11. GCC 10 and earlier will appear in subsequent days.

As each are released, the exact `config.mak` files I used will also be added to https://conf.musl.cc/ (you'll want `plain_*.txt`).

**Reply by userdocs** [reply]
I think avoiding using my builds is the best option but using the workflow modified to fall inline with what @zv-io releases (for gcc 11) over at musl.cc, where the config is for those releases is documented, is a good idea.

This way your config can be inline with the stable releases but adaptable to any specific needs you have for building this app and you can have those locally available to you and tailored to your build process for. For example, how the build process is documented and tools deployed to the end user.

Now it's easy to build on github and easy to provide a documented way to build on the end users system using a tailored cross build package. Even via some docker commands.

I think that' the best route.

**Reply by zv-io** [reply]
@userdocs it's been on my TODO list /forever/ to set up automatic build/test infrastructure on my end, which will make it easy to plumb in packaging (e.g. Docker, `.deb`, `.rpm`, etc.). I would be doing this on GitLab, though. Would you have any interest in helping with that?

**Reply by userdocs** [reply]
The only issue is I have never used Gitlab CI and one reason is that it's quite different from the github workflow syntax. Also there are many actions I use on github for creating debs and stuff. So I have no idea how this is done on GitLab and would need to learn.

I am open minded to trying but I guess I'll have to start with a basic goal and go figure it out over there. Perhaps like exporting the current workflow to GitLab and getting it to work and provide an identical outcome

**Reply by zv-io** [reply]
OK. For now, I'll work on this next release and poke at automation later. Are you in IRC?

**Reply by userdocs** [reply]
I am loitering in a few libera.chat channels or i use discord (which i prefer) or even matrix (via element)

**Reply by jan-guenter** [reply]
@zv-io @userdocs I work in DevOps now (former c++ dev) and have been working with an on-premise GitLab in our company since about 6 years now. If you like my help in setting up the automation feel free to contact me. @zv-io I also did the setup of our CI network and can lend a hand with the gitlab-runner setup for your instance.
My usual communication channels are [email](mailto:jan.guenter@googlemail.com) or [Discord](https://discordapp.com/users/jellow#9966), but I can also hop on an IRC if you like.
